.github/workflows/fuzzy-ci.yml

```diff
@@ -0,0 +1,475 @@
+            echo "--------end of full responses diff head--------"
+            echo "--------beginning of distilled data diff head--------"
+            head -n 100 "$DISTILLED_DIFF_FILE"
+            echo "--------end of distilled data diff head--------"
+          }
+
+          # FIXME (?): Are nested conditionals always so ugly in Bash, or is there a better way? Option types and the possibility to match would help a lot.
+          LABEL_NAME=$(cat .github/fuzzy-ci-helpers/label_name.txt)
+          if $earlier_diff_was_approved; then
+            echo "Earlier diff was approved."
+            if [ ${{ steps.current_diff.outputs.hash }} == ${{ steps.approved_diff.outputs.hash }} ]; then
+              echo "This diff has been approved earlier. Everything ok."
+              exit 0
+            else
+              print_head_of_diffs
+              printf "The diff has changed since it was approved. So I'm removing the $LABEL_NAME label. If the new diff looks good, please set the label again.\n\
+              There's a head of the new diffs printed above. The whole diffs can be downloaded from $CURRENT_ACTION_URL .\n\
+              Previous sha256: ${{ steps.approved_diff.outputs.hash }}\n\
+              Current sha256: ${{ steps.current_diff.outputs.hash }}"
+              echo "delete_label=true" >> $GITHUB_OUTPUT
+              exit 1
+            fi
+          else
+            if $current_diff_exists; then
+              print_head_of_diffs
+              printf "There's a head of the diffs printed above. The diffs can be downloaded from $CURRENT_ACTION_URL .\nIf it looks good, please set the $LABEL_NAME label on the PR."
+              exit 1
+            else
+              echo "No diff. All good."
+              exit 0
+            fi
+          fi
+
+
+  approve:
+    name: Approve diff
+    if: >
+      github.event_name == 'pull_request' &&
+      github.event.action == 'labeled' &&
+      github.event.label.name == 'fuzzy-diff-looks-good'
+    runs-on: ubuntu-22.04
+    steps:
+      - name: Retreive diff artifact meta-data
+        id: diff_metadata
+        run: |
+          all_artifacts=$(curl -sSL  "$GH_API_ARTIFACTS")
+          diff_artifact=$(echo "$all_artifacts" | jq "first(.artifacts[] | select(.name == \"$DIFF_ARTIFACT_NAME\")  )")
+          id=$(echo "$diff_artifact" | jq ".id")
+          echo "id=$id" | tee -a $GITHUB_OUTPUT
+          workflow_run=$(echo "$diff_artifact" | jq ".workflow_run | .id")
+          echo "workflow_run=$workflow_run" | tee -a $GITHUB_OUTPUT
+          if [ -z $id ]; then
+            echo "exists=false" | tee -a $GITHUB_OUTPUT
+          else
+            echo "exists=true" | tee -a $GITHUB_OUTPUT
+          fi
+
+      - name: Write instruction to delete PR label
+        # When this workflow is triggered by a PR from a fork, it doesn't have
+        # the permissions to delete PR labels. Instead, we forward the
+        # instruction to delete the label to fuzzy-ci-privileged.yml.
```

Comment:
That's so convoluted^^

(this is not a comment asking for changes 🙂)

Reply:
Yes :100: :sweat_smile: 

However, in general, they have good reason to make it that complicated. Workflows triggered by PRs can checkout the code on the PR source and build it, and the PR can alter the build binaries and the workflow behavior. So I understand that they strongly limit the permissions of workflows triggered by PRs. There's a [series of blog posts](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/) on securitylab.github.com about that.

I personally think with more granularity on the permissions, the actions I need for this workflow (just commenting on the PR and deleting a label on the PR) shouldn't be a huge deal for security, but the general point they have is a good one.